### PR TITLE
Implementing component-handles-ref and component-has-root-ref tests within react-conformance

### LINKED ISSUE #15334: Add ref conformance test
To test all our FCs can receive `ref` and `ref` is correctly forwarded to root element of the component.

### PR DESCRIPTION
**Pull request by czearing**
Implementing component-handles-ref and component-has-root-ref tests within react-conformance

#### Pull request checklist
- [X] Addresses an existing issue: Fixes #15334
- [X] Include a change request file using `$ yarn change`

#### Description of changes

Introducing two new react-conformance tests:

#### `component-handles-ref`

Component handles ref.
```jsx
  'component-handles-ref': (componentInfo: ComponentDoc, testInfo: IsConformantOptions) => {
    it(`handles ref`, () => {
      const { Component, requiredProps, elementRefName = 'ref' } = testInfo;
        const rootRef = React.createRef<HTMLDivElement>();
        const mergedProps: Partial<{}> = {
          ...requiredProps,
          [elementRefName]: rootRef,
        };

        act(() => {
          mount(<Component {...mergedProps} />);
          expect(rootRef.current).toBeDefined();
          // Ref should resolve to an HTML element.
          expect(rootRef.current?.getAttribute).toBeDefined();
        });
    });
  },
  ```
  
  **defaultErrorMessages**
  
  
  
#### `component-has-root-ref`

Component has ref applied to the root component DOM node.

```jsx
  'component-has-root-ref': (componentInfo: ComponentDoc, testInfo: IsConformantOptions) => {
    it(`ref exists on root element`, () => {
      const {
        customMount = mount,
        Component,
        requiredProps,
        helperComponents = [],
        wrapperComponent,
        elementRefName = 'ref',
        passesUnhandledPropsTo,
      } = testInfo;

      const rootRef = React.createRef<HTMLDivElement>();
      const mergedProps: Partial<{}> = {
        ...requiredProps,
        [elementRefName]: rootRef,
      };

      const el = passesUnhandledPropsTo
        ? customMount(<Component {...mergedProps} />).find(passesUnhandledPropsTo)
        : customMount(<Component {...mergedProps} />);

        act(() => {
          const component = getComponent(el, helperComponents, wrapperComponent);

          expect(rootRef.current).toBe(component.getDOMNode());
        });
    });
  },
```

**defaultErrorMessages**
  
Example of diffing between ref.current and DOM node.

Example of ref.current returning null instead of the DOMD node.

#### Disabled Components
This test brought attention to several components that don't handle ref properly, being that they either don't pass to the component's root or don't contain a ref.

(Here's the details why these components had the tests disabled and future solutions [hackMD](https://hackmd.io/tkGXMpeoRya5N7xWDdlufw?both))

#### `date-time`

WeeklyDayPicker: `component-has-root-ref`

#### `react`

Breadcrumb `component-has-root-ref` `component-handles-ref`
ComboBox `component-has-root-ref` `component-handles-ref`
GroupedList `component-has-root-ref` `component-handles-ref`

#### `react-internal`

**Requires Work Around:**
ComboBox `component-has-root-ref` `component-handles-ref`
Modal `component-has-root-ref` `component-handles-ref`

**Ref isn't passed:**
SpinButton `component-has-root-ref` `component-handles-ref`
ContextualMenu `component-has-root-ref` `component-handles-ref`
Breadcrumb `component-has-root-ref` `component-handles-ref`
Text `component-has-root-ref` `component-handles-ref`
Stack `component-has-root-ref`
ProgressIndicator `component-has-root-ref` `component-handles-ref`
Overlay `component-has-root-ref` `component-handles-ref`
Nav `component-has-root-ref` `component-handles-ref`
Icon `component-has-root-ref`
Dialog `component-has-root-ref` `component-handles-ref`
Label `component-has-root-ref` `component-handles-ref`
Spinner `component-has-root-ref` `component-handles-ref`
ActivityItem `component-has-root-ref` `component-handles-ref`
FacePile `component-has-root-ref` `component-handles-ref`
Check `component-has-root-ref`
Announced `component-has-root-ref` `component-handles-ref`
CommandBar `component-has-root-ref` `component-handles-ref`
Separator `component-has-root-ref`
SelectionZone `component-has-root-ref` `component-handles-ref`
ColorPicker `component-has-root-ref` `component-handles-ref`
PeoplePicker `component-has-root-ref` `component-handles-ref`
TagPicker `component-has-root-ref` `component-handles-ref`
BaseSelectedItemsList `component-has-root-ref` `component-handles-ref`
SelectedItemsList `component-has-root-ref` `component-handles-ref`

**Ref is not applied to the root DOM node:**
~~TextField `component-has-root-ref`~~
MarqueeSelection `component-has-root-ref` `component-handles-ref`
HoverCard `component-has-root-ref` `component-handles-ref`
Panel `component-has-root-ref`
List `component-has-root-ref` `component-handles-ref`
Sticky `component-has-root-ref`
BasePicker `component-has-root-ref` `component-handles-ref`
Suggestion `component-has-root-ref` `component-handles-ref`
Callout `component-has-root-ref`
Image `component-has-root-ref`
TeachingBubble `component-has-root-ref`
VerticalDivider `component-has-root-ref` `component-handles-ref`

##### `react-northstar`

Planned to be enabled in a future pull request.
Here are all of the failed components: [hackMD](https://hackmd.io/OAUn0pF6Qj-vc315wAHXLQ)

##### `date-time`
* **Bug Fix** Changed CalendarDayGrid's FocusZone ref to an elementRef.

##### `react-conformance`
* **Type Rename** Renamed passesUnhandledProps to targetComponent
* **API Change** Added elementRefName type to allow for elementRef's to use this test.

### CODE REVIEW

File: packages/react-internal/src/components/ContextualMenu/ContextualMenu.test.tsx [hunk 1]
```diff
@@ -31,6 +31,9 @@ describe('ContextualMenu', () => {
   isConformant({
     Component: ContextualMenu,
     displayName: 'ContextualMenu',
+    // Problem: Returns FunctionComponents can’t be given refs
+    // Solution: Missing a forwardRef:
```
Comment (xugao) [inline packages/react-internal/src/components/ContextualMenu/ContextualMenu.test.tsx:35]:
```suggestion
    // Problem: Is FunctionComponent but missing `forwardRef`
    // Solution: add forwardRef and ref typing
```

Comment (xugao) [reply packages/react-internal/src/components/ContextualMenu/ContextualMenu.test.tsx:35]: let's get this fixed before v8 beta

File: packages/react-internal/src/components/Check/Check.test.tsx [hunk 2]
```diff
@@ -5,5 +5,8 @@ describe('Check', () => {
   isConformant({
     Component: Check,
     displayName: 'Check',
+    // Problem: Ref isn't passed.
+    // Solution: Ref should be added and passed onto the root.
```
Comment (xugao) [inline packages/react-internal/src/components/Check/Check.test.tsx:9]:
```suggestion
    // Problem: Is FunctionComponent but missing `forwardRef`
    // Solution: add forwardRef and ref typing
```

Comment (xugao) [reply packages/react-internal/src/components/Check/Check.test.tsx:9]: Let's get this fixed before v8 beta

File: packages/react-conformance/src/defaultTests.tsx [hunk 3]
```diff
@@ -82,6 +82,62 @@ export const defaultTests: TestObject = {
...
+        const mergedProps: Partial<{}> = {
+          ...requiredProps,
+          [elementRefName]: rootRef,
+        };
+
+        customMount(<Component {...mergedProps} />);
+
+        expect(rootRef.current).toBeDefined;
+        // Ref should resolve to an HTML element.
+        expect(rootRef.current?.getAttribute).toBeDefined;
```
Comment (xugao) [inline packages/react-conformance/src/defaultTests.tsx:100]:
```suggestion
        expect(rootRef.current).toBeDefined();
        // Ref should resolve to an HTML element.
        expect(rootRef.current?.getAttribute).toBeDefined();
```

Comment (xugao) [reply packages/react-conformance/src/defaultTests.tsx:100]: I was wondering why this was passing for some tests; a lot more tests will need to disable `component-handles-ref`

File: packages/react-internal/src/components/Dialog/Dialog.test.tsx [hunk 4]
```diff
@@ -64,6 +64,9 @@ describe('Dialog', () => {
   isConformant({
     Component: Dialog,
     displayName: 'Dialog',
+    // Problem: Doesn’t apply ref to the root element.
+    // Solution: Ensure that the ref is applied to the root element.
```
Comment (xugao) [inline packages/react-internal/src/components/Dialog/Dialog.test.tsx:68]:
```suggestion
    // Problem: ref is not supported
    // Solution: Convert to FunctionComponent and support using forwardRef
```

Comment (xugao) [reply packages/react-internal/src/components/Dialog/Dialog.test.tsx:68]: pls apply similar comments to other places where we `ref` is not/never supported due to we have not converted them to FC yet

File: packages/react-internal/src/components/Image/Image.test.tsx [hunk 5]
```diff
@@ -27,6 +27,9 @@ describe('Image', () => {
   isConformant({
     Component: Image,
     displayName: 'Image',
+    //Problem: Not applied to the root properly
```
Comment (xugao) [inline packages/react-internal/src/components/Image/Image.test.tsx:30]:
```suggestion
    // Problem: Ref is applied but not on root element
```

File: packages/react/src/components/ComboBox/ComboBox.test.tsx [hunk 6]
```diff
@@ -76,8 +76,10 @@ describe('ComboBox', () => {
   isConformant({
     Component: ComboBox,
     displayName: 'ComboBox',
-    // Disabled due to being required to mount by safeCreate. A test called "renders" was added here as a replacement.
-    disabledTests: ['component-renders'],
+    // Problem: Currently doesn’t work with mount.
+    // Solution: Further investigate why ComboBox only works with create or fix customMount type
```
Comment (xugao) [inline packages/react/src/components/ComboBox/ComboBox.test.tsx:80]: what's the error you are getting?
(note: this is only a test issue. ref is working correctly)

**Review by xugao — CHANGES_REQUESTED** [review]

File: packages/react/src/components/ComboBox/ComboBox.test.tsx (hunk 6 above)
Comment (czearing) [reply packages/react/src/components/ComboBox/ComboBox.test.tsx:80]: When ComboBox uses enzyme mount it returns `Cannot read property 'tag' of null`. I'm not certain of the reason behind this but it seems to work correctly when using create.

Comment (xugao) [reply packages/react/src/components/ComboBox/ComboBox.test.tsx:80]: oh i know why. see this change I have [here](https://github.com/microsoft/fluentui/pull/15534/files#diff-4f25026635039c045750d43d45e8e7ca20b1141746eeafdaab93a93fde2e2cdd). `mount` relays on `createPortal` but combobox tests are mocking the `createPortal`. if you reset `createPortal` to original implementation before running the conformance tests, that should help resolve

File: packages/react-internal/src/components/ContextualMenu/ContextualMenu.test.tsx [hunk 7]
```diff
@@ -31,6 +31,10 @@ describe('ContextualMenu', () => {
   isConformant({
     Component: ContextualMenu,
     displayName: 'ContextualMenu',
+    // Priority: High
+    // Problem: Is FunctionComponent but missing `forwardRef`
+    // Solution: add forwardRef and ref typing
+    disabledTests: ['component-has-root-ref', 'component-handles-ref', 'component-handles-ref'],
```
Comment (xugao) [inline packages/react-internal/src/components/ContextualMenu/ContextualMenu.test.tsx:37]:
```suggestion
    disabledTests: ['component-has-root-ref', 'component-handles-ref'],
```

**Review by xugao — APPROVED** [review]

File: packages/react-conformance/src/types.ts [hunk 8]
```diff
@@ -65,16 +65,20 @@ export interface IsConformantOptions<TProps = {}> {
...
   /**
    * If the component's 'as' property requires a ref, this will attach a forwardRef to the test component passed to 'as'
    * and disable the as-renders-react-class test.
    */
   asPropHandlesRef?: boolean;
+  /**
+   * An alternative name for the ref prop which resolves to
+   * the root element. (E.g. "elementRef") Defaults to 'ref'.
+   */
+  elementRefName?: string;
```
Comment (ecraig12345) [inline packages/react-conformance/src/types.ts:77]:
```suggestion
  /**
   * An alternative name for the ref prop which resolves to
   * the root element (e.g. `elementRef`).
   * @defaultvalue 'ref'
   */
  elementRefName?: string;
```

**Review by ecraig12345 — APPROVED** [review]
Mostly looks good, thanks! One super minor comment.

**Pull request merged**